### DISCUSSION TITLE
How do I make nvim-cmp show the completions even if there's whitespace before the cursor?

### DISCUSSION THREAD

**Discussion by luiz00martins** [hrsh7th/nvim-cmp#1353] [Q&A] [answered]
By default nvim-cmp does not show the completions if there is whitespace before the cursor, for example:
```
def something|
def something |
```
The first line will show the completions, the second one won't.

How may I make nvim-cmp always show the completions, even if there's whitespace before the cursor?

---

Update: I actually got a response for one of my sources:

![image](https://user-images.githubusercontent.com/43142209/208076390-e517d8ba-42c2-4029-a1da-a4d89e206f33.png)

But only this one. I think this might a source problem then.

**Comment by fitrh** [comment, marked as answer]
**UPDATE**

You can create autocmd

```lua
vim.api.nvim_create_autocmd({ "CursorHoldI", "TextChangedI" }, {
  group = vim.api.nvim_create_augroup("cmp_complete_on_space", {}),
  callback = function()
    local line = vim.api.nvim_get_current_line()
    local cursor = vim.api.nvim_win_get_cursor(0)[2]

    if string.sub(line, cursor, cursor + 1) == " " then
      require("cmp").complete()
    end
  end,
})
```

---

Set a keybind for `cmp.complete()`, something like this

```lua
 mapping = cmp.mapping.preset.insert({
     ['<C-l>'] = cmp.mapping.complete(),
 }),
```

**Reply by luiz00martins** [reply]
I mean automatically. I already have a keybinding, but pressing it everytime is not ideal.

**Reply by fitrh** [reply]
You can create autocmd

```lua
vim.api.nvim_create_autocmd("CursorHoldI", {
  group = vim.api.nvim_create_augroup("cmp_complete_on_space", {}),
  callback = function()
    local line = vim.api.nvim_get_current_line()
    local cursor = vim.api.nvim_win_get_cursor(0)[2]

    if string.sub(line, cursor, cursor + 1) == " " then
      require("cmp").complete()
    end
  end,
})
```

**Reply by luiz00martins** [reply]
This actually works:

![image](https://user-images.githubusercontent.com/43142209/208081249-f39fe5d9-95d3-4232-8684-19e3bb0b6f13.png)


However, it only pops after 4 seconds. (You could change the `updatetime`, but that would mean putting you swap-file-saving to overdrive)

**Reply by fitrh** [reply]
Then you can use `TextChangedI` event if you don't want to depend on `'updatetime'`, although I find that is too aggressive

**Reply by fitrh** [reply]
> But only this one. I think this might a source problem then.

No, it is not a source problem, what do you get is the response of the lua language server when you use `type` function (or any expression that expects you to use enum-like values)

**Reply by luiz00martins** [reply]
> what do you get is the response of the lua language server when you use type function

That's why I said it's source problem. It worked with the lua language server, but not with tabnine. Therefore, the problem is with the tabnine source. And I can confirm that: https://github.com/tzachar/cmp-tabnine/issues/83

**Comment by par4m** [comment]
I'm trying to do the opposite, any idea how? I want to make Tab and S-Tab function normally but if there's a character and not a whitespace they should cycle the menu.

**Reply by fitrh** [reply]
> but if there's a character and not a whitespace they should cycle the menu.

Can you explain how this looks like?
